### ISSUE TITLE
@babel/preset-typescript parsing errors

### ISSUE THREAD

**Issue by drcmda** [babel/babel#6768]
Choose one: is this a bug report or feature request? 

Bug

I've basically added the typescript preset without making any changes, the code is not typed. It seems to hang up on a few things it has trouble parsing.

### Input Code

```js
const arg = undefined
!!arg && arg.foo()
```

Will cause parsing issues, like:

```js
ERROR in ../awv3/three/object3.js
Module build failed: SyntaxError: object3.js: Unexpected token, expected ; (233:10)
  231 |         await this.emit(Object3.Events.Lifecycle.ChildAdded, { child: object })
  232 |         await object.emit(Object3.Events.Lifecycle.Added, { parent: this })
> 233 |         !!this.view && this.view.invalidate()
      |           ^
  234 |     }
  235 | }

ERROR in ../awv3/session/selector.js
Module build failed: SyntaxError: selector.js: Unexpected token, expected ; (91:17)
  89 |             this.on(['add', 'remove', 'removeAll'], () => {
  90 |                 const state = this.selectedIds
> 91 |                 !isEqual(state, oldState) && this.emit('changed', state, oldState)
```

And a separate issue. When i use [babel-plugin-import](https://github.com/ant-design/babel-plugin-import) which rewrites imports.

```js
import React from 'react'
import { Icon } from 'antd'

class Test extends React.Component {
    render() {
        return <Icon type="cloud-download" />
    }
}
```

causes

```js
ERROR in ./components/RunChecks.js
Module build failed: TypeError: Cannot read property 'referencePaths' of undefined
    at isImportTypeOnly (node_modules/@babel/plugin-transform-typescript/lib/index.js:232:35)
```

### Babel/Babylon Configuration (.babelrc, package.json, cli command)

```js
const babelOptions = {
    babelrc: false,
    cacheDirectory: true,
    presets: [
        ['@babel/preset-env', { modules: false, loose: true, targets: { chrome: 60 } }],
        '@babel/preset-stage-0',
        '@babel/preset-react',
        '@babel/preset-typescript',
    ],
    plugins: [
        ['babel-plugin-import', { libraryName: 'antd', style: 'css' }],
    ],
}
```

### Your Environment

```js
    "@babel/core": "7.0.0-beta.31",
    "@babel/preset-env": "7.0.0-beta.31",
    "@babel/preset-react": "7.0.0-beta.31",
    "@babel/preset-stage-0": "7.0.0-beta.31",
    "@babel/preset-typescript": "^7.0.0-beta.31",
    "babel-loader": "8.0.0-beta.0",
    "babel-plugin-import": "1.6.2",
    "antd": "3.0.0-beta.6",
```

| software         | version(s)
| ---------------- | -------
| node             | 8.9.9
| npm              | 5.5.1
| Operating System | High Sierra

**Comment by hzoo** [comment]
Cool, can you provide the minimal snippet of code that actually fails (not just the code frame)? Then we can repo.

**Comment by drcmda** [comment]
sure, this is enough to bring it down:

```js
const view = undefined
!!view && view.foo()
```

The `Cannot read property 'referencePaths' of undefined` error is an internal one (plugin-transform-typescript/lib/index.js:232:35), i wouldn't know how to trigger it.

**Comment by existentialism** [comment]
Related (transform-typescript + transform-runtime): https://github.com/babel/babel/issues/6093#issuecomment-322944634

**Comment by hzoo** [comment]
@drcmda I don't see 4 pkgs in the config, so you can remove those?

    "@babel/plugin-proposal-class-properties": "7.0.0-beta.31",
    "@babel/plugin-proposal-decorators": "7.0.0-beta.31",
    "@babel/plugin-transform-classes": "7.0.0-beta.31",
    "@babel/plugin-transform-runtime": "7.0.0-beta.31",

**Comment by hzoo** [comment]
Sounds like a parser issue with having no semicolons, it works fine without the typescript preset?

**Comment by drcmda** [comment]
@hzoo Yes, without the preset it runs fine.

**Comment by drcmda** [comment]
The other issue seems to be related to the [babel-plugin-import](https://github.com/ant-design/babel-plugin-import) plugin

```js
    plugins: [
        'react-hot-loader/babel',
        ['babel-plugin-import', { libraryName: 'antd', style: 'css' }],
    ],
```

```js
import React from 'react'
import { Icon } from 'antd'

export default class EmptyCanvas extends React.Component {
    render() {
        return this.props.connections.length > 0 ? null : (
            <div>
                <Icon type="cloud-download" style={{  fontSize: 100, display: 'inline' }} />
                <span style={{ marginTop: 5, fontSize: 18 }}>Drop your files</ span>
            </div>
        )
    }
}
```

This plugin will rewrite the import statement.

**Comment by hzoo** [comment]
Ok 😂 then I'm never heard of that plugin and it's not one of our supported ones? Seems like it needs to be updated for v7 (most likely) or typescript plugin needs an update?

**Comment by drcmda** [comment]
I think it's more or less the same issue as in https://github.com/babel/babel/issues/6093#issuecomment-322944634 - the moment any plugin touches imports typescript crashes.

This was a separate issue though. Comma parsing crashes regardless.

**Comment by Lodin** [comment]
I have the same problem as @drcmda.

I'm working on custom react-typescript boilerplate project ([link](https://github.com/Lodin/poetez-ts-reactapp-boilerplate)), and recently I've started moving it to babel 7 stack instead of `tsc`. Problems started when I tried to execute `npm run build` script (`npm start` works well, not sure why).

Error stack leads to [isImportTypeOnly](https://github.com/babel/babel/blob/master/packages/babel-plugin-transform-typescript/src/index.js#L243): it complains that `binding` argument is undefined for some reason. 

I use `@babel/plugin-transform-runtime`, and if I remove it from plugins list, everything gets back to normal.
